### PR TITLE
[#7] Configurable retries to close/release requests

### LINKED ISSUE #7: Add configurable retries to all requests
close and release operations in Nexus can take some time. It is required to wait a little bit until the command is effectively performed (i.e. is not in the transitioning state) after it was accepted.

### PR DESCRIPTION
**Pull request by szpak**
his PR adds retrying on the close and release requests. It is required as Nexus only acknowledge that a given repository has been accepted for transition. With Sonatype Nexus is can take minutes to finish the operation.

I started with a basic external [retrier](https://github.com/alexo/retrier), but separated it to be able to switch to some other (external or in-house implementation if needed or desired).

I also considered this [retrier](https://github.com/ravichaturvedi/retrier) (it seems to only support throwing exceptions to indicate a not successful attempt) and that one for [kotlin](https://github.com/michaelbull/kotlin-retry) (it uses suspended function and I'm not sure how Gradle deals with that in the plugins and in addition it is not in Maven Central nor jCenter). In GNSP, I have a custom implementation (which among others nicely displays subsequent attempts), but I don't know if we want to reinvent a wheel here and maybe some external retrier will be fine.

I was able to successfully release locally a small sample project with the plugin as e2e tests. However, it requires some more work to be able to use it as automatic e2e tests from the CI server (#5). Then it would be good to split also functional tests into smaller pieces.

Btw, I moved MockServer to a field to simplify private method calls - it is used by (almost) everything in that class. It makes a diff less readable. I may revert it and propose as a separate PR, if you prefer.

Closes #7.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -110,7 +115,9 @@ configurations {
 sourceSets {
     compatTest {
         compileClasspath += sourceSets["test"].output
+        compileClasspath += sourceSets["main"].output
```
Comment (vlsi) [inline build.gradle.kts]: Do you mean `.output.classesDirs` here?

File: src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.github.gradlenexus.publishplugin.internal
+
+interface ActionRetrier<R> {
+    fun execute(operationToExecuteWithRetrying: () -> R): R
+}
```
Comment (vlsi) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt]: Kotlin does not support implementation of `Kotlin` interfaces via lambdas.

Do you think it would be better to use `typealias ActionRetrier`?

Comment (vlsi) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt]: Ah, `typealias` can't have generics :(

Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt]: > Kotlin does not support implementation of `Kotlin` interfaces via lambdas.

I didn't know. I could change it to Java interface if that helped.

> Do you think it would be better to use `typealias ActionRetrier`?

`typealias` seems to be more an a type alias (like in Groovy or Scala) and I would like to have an interface with (potentially) multiple implementations.

File: build.gradle.kts (hunk 1 above)
Comment (szpak) [reply build.gradle.kts]: The above works. Probably there is an overloaded method to support it.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt (hunk 2 above)
Comment (vlsi) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt]: >I could change it to Java interface if that helped.

In fact, it would help. In other words, Kotlin can implement Java interfaces via lambdas.
However, it is a bit weird for me :(

File: build.gradle.kts (hunk 1 above)
Comment (vlsi) [reply build.gradle.kts]: `.output` is more than `classes`. it includes resources as well.

Do you really need to include `resources` on the **compile**  classpath?

Comment (szpak) [reply build.gradle.kts]: To compile not, but in runtime it might be necessary in some cases. E.g. with Nebula Test it would be required to provide a properties file with the main plugin class provided (with a plugin for plugins and TestKit is not necessary, however).

File: build.gradle.kts [hunk 3]
```diff
@@ -80,6 +80,9 @@ configurations {
 dependencies {
     shadowed("com.squareup.retrofit2:retrofit:2.7.1")
     shadowed("com.squareup.retrofit2:converter-gson:2.7.1")
+    shadowed("io.github.alexo:retrier:0.1") {
+        exclude("org.mockito")  // https://github.com/alexo/retrier/pull/1
+    }
```
Comment (marcphilipp) [inline build.gradle.kts]: I've had a good experience with https://jodah.net/failsafe/ for this in the past. It seems more mature, too. Should we switch?

File: src/main/kotlin/io/github/gradlenexus/publishplugin/NexusRepository.kt [hunk 4]
```diff
@@ -48,6 +52,18 @@ open class NexusRepository @Inject constructor(@Input val name: String, project:
...
 
+    @Nested
+    val retrying = project.objects.property<RetryingConfig>().apply {
+        set(RetryingConfig(project))
+    }
+
+    fun retrying(action: Action<in RetryingConfig>) = action.execute(retrying.get())
+
+    //TODO: Fails anyway with Groovy for "maxNumber = 10" with "Cannot cast object '10' with class 'java.lang.Integer' to class 'org.gradle.api.provider.Property'"
+    fun retrying(closure: Closure<*>) {
```
Comment (marcphilipp) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/NexusRepository.kt]: :x: We don't need this method because you can write an `Action` as a closure in Groovy.

Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/NexusRepository.kt]: I believe that there was something mentioned in the Gradle documentation that Gradle doesn't support the second level of nesting domain objects and it is needed. Unfortunately, I cannot find it right now. Nevertheless, it doesn't work as mentioned in the comment for Groovy anyway, so I maybe remove it anyway.

File: build.gradle.kts (hunk 3 above)
Comment (szpak) [reply build.gradle.kts]: I didn't know it. I will take a look.

Comment (szpak) [reply build.gradle.kts]: It's very nice. That's pity that is was in my search results at the end of the second page :-/. Replaced.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/internal/BasicActionRetrier.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import io.github.gradlenexus.publishplugin.StagingRepository
+import net.jodah.failsafe.Failsafe
+import net.jodah.failsafe.RetryPolicy
+
+open class BasicActionRetrier<R>(retryingConfig: RetryingConfig, stopFunction: (R) -> Boolean) : ActionRetrier<R> {
+
+    private val retrier: RetryPolicy<R> = RetryPolicy<R>()
+            .handleResultIf(stopFunction)
+            .withMaxRetries(retryingConfig.maxRetries.get())
+            .withDelay(retryingConfig.delayBetween.get())
```
Comment (szpak) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/internal/BasicActionRetrier.kt]: Failsafe is nice and probably well tested, but I would like to test that I pass the parameters correctly. Functional tests check retrying in general (if it occurs), but I would like to test some corner cases (such as a reaction on particular exceptions) at the unit (testing) layer.
Unfortunately, `RetryingConfig` relies on the Gradle API (for `Property`). I initially modeled it as a implementation of a generic interface which just provides values, but it complicated the code. Do you have any better/simpler idea how would we make `BasicActionRetries` less Gradle dependent?

Comment (marcphilipp) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/internal/BasicActionRetrier.kt]: I think I would pass in `retryingConfig.maxRetries.get()` and `retryingConfig.delayBetween.get()` to the constructor instead of `RetryingConfig`.

Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/internal/BasicActionRetrier.kt]: That's definitely much simpler solution :-). Probably we will not have (much) more parameters passes here in the future.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt (hunk 2 above)
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/internal/ActionRetrier.kt]: I gave it a try, but it complicates the current implementation a little bit.

```
interface ActionRetrier<R> {
    R execute(Supplier<R> operationToExecuteWithRetrying);
}
```
cannot be used as FailSafe which we use expects its own:
```
public interface CheckedSupplier<T> {
  T get() throws Exception;
}
```
Callable would be ok, but at the type level they are not compatible. As a result:
```
    override fun execute(operationToExecuteWithRetrying: Callable<R>): R {
        return Failsafe.with(retrier).get(
            { operationToExecuteWithRetrying.call() } as CheckedSupplier<R>) }
        )
```
is required which doesn't look good. 

I will merge it as is (with the Kotlin interface). Feel free to create a separate PR to improve that.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/RetryingConfig.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Internal
+    val maxRetries = project.objects.property<Int>().apply {
+        set(DEFAULT_MAXIMUM_NUMBER_OF_RETRIES)
+    }
+
+    @Internal
+    val delayBetween = project.objects.property<Duration>().apply {
+        set(DEFAULT_DELAY_BETWEEN_RETRIES)
+    }
```
Comment (vlsi) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/RetryingConfig.kt]: The naming is misleading.
`RetryingConfig` + `maxRetries` means the plugin would **retry** publication multiple times, however, in practice it means it would "just wait the specified number of times while the repository is in progress".

Do you think it would be better to use `timeout`-like configuration properties?

For instance `connectToNexusTimeout`, `initializeRepositoryTimeout`, `singleArtifactPublishTimeout`, `closeRepositoryTimeout`, etc, etc.

Note: the users do not care what the number of "status check" attempts is there. What they care is **ovreall timeout value**, and they might want to have something like `initial delay between status check attempts`.


WDYT?

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt [hunk 7]
```diff
@@ -497,6 +501,7 @@ class NexusPublishPluginTests {
     }
 
     @Test
+    @Disabled("Fails on my Fedora...")
```
Comment (szpak) [inline src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt]: Ups, I have to do some conditional checking to exclude only my system as it's hard to keep it locally and do not commit accidentally...
